Object coordinates: It is now possible to hear object coordinates when using object navigation commands. re #2559

josephsl opened the pull request:
Issue 2559: Sometimes, it might be desirable to get an idea as to where a given control is located on screen. However, the only way to do this is mouse tracking, which has as serious design flaw in that it cannot be extended to cover situations where obj nav could trigger coordinate beep playback.

The following changes were made:
- Screen explorer: Adds a new function to play obj coordinates given the object in question.
- Mouse handler: Object coordinate announcement will not use the new function introduced in screen explorer.
- Global commands: Next/previous/parent/firstChild commands and their touch equivalents will play obj coordinates if told to do so.
- Object presentation dialog: Added a checkbox to play object coordinates, now documented this in the user guide.

Thanks.

josephsl committed:
Object coordinates: It is now possible to hear object coordinates when using object navigation commands (rebased with master in June 2016). re #2559

michaelDCurran commented on source/screenExplorer.py:
```diff
@@ -1,8 +1,59 @@
+	l,t,w,h=obj.location
+	x = l+(w/2)
+	y = t+(h/2)
+	screenWidth, screenHeight = api.getDesktopObject().location[2], api.getDesktopObject().location[3]
```
Rather than calling getDesktopObject twice,  just slice the location:
screenWidth,screenHeight=api.getDesktopObject().location[2:]

michaelDCurran commented:
Assuming this will not be used for focus, I think the user-visible option should really be called something like Play object coordinates when navigating or some such. Otherwise  it looks like it may apply to all objects.

josephsl commented:
Hi,
Ah, thanks for catching that... Will apply it (perhaps it should be really be called "Play object coordinates when moving between objects" or a clearer label...). Thanks.

josephsl commented:
Hi,

potential bug found: if #5906 shows up on master, it'll mean one of the new beeps (indentation or obj coordinate) will not be played.
@derekriemer and @MichaelDCurran: what do you suggest? Thanks.

michaelDCurran commented:
@josephsl: I think you have 1 or two points to address on this pr before it can be incubated.

josephsl commented:
Hi,

Based on discussions between @jcsteh, @MichaelDCurran and I, it was decided to let become nav object event fire this event so it can apply everywhere. Along with slicing fix, it'll be done sometime tomorrow. Thanks.

josephsl commented:
Hi,

Besides the ones listed above, are there other things that should be addressed? Thanks.

josephsl committed:
NVDAObjects: obj coordinate announcement will be handled via becomeNavObj event. re #2559

As object coordinate announcement will be done regardless of focus movement, it was decided to let becomeNavObject event call call object coordinate player function in screen explorer. This means global commands will not use this method (this decisoin was made after talking to Jamie Teh and others).

michaelDCurran commented:
@josephsl can you merge master again and fix conflicts? Sorry,  I left this for too long.

josephsl committed:
Object coordinates: obj coordinates checkbox now follows guiHelper services. re #2559

josephsl committed:
Object coordinates: updated copyright headers, removed unused imports.

josephsl committed:
Config: play obj coordinates validator now lives in the new config spec module (after rebasing with master). re #2559.

josephsl committed:
Merged master

josephsl committed:
Object coordinates: applied multi-monitor fix from NVDA 2017.1. re #2559.

In NVDA 2017.1, it is possible to hear mouse coordinates on multi-monitor systems. Thus object coordinate announcement function has been modified to take advantage of this. To ensure backward compatibility, minimum screen position paremeter will be a keyword argument, ready to become mandatory if obj coordinate announcement from multi-monitor set ups is desired.

josephsl committed:
Merge branch 'master' into i2559

# Conflicts:
#	source/NVDAObjects/__init__.py

josephsl committed:
Object coordinates: do not play the tone if the object ia offscreen. re #2559

leonardder commented:
@josephsl: This has merge conflicts again. If fixed, I hope @michaeldcurran will be able to address this before 2018.1?

josephsl commented:
Hi, thinking about it, I’d say we still need to deal with conflicts with indentation tones, but other than that, I’d be happy to target 2018.1. Extension points may or may not help in this situation. Thanks.

 

From: Leonard de Ruijter [mailto:notifications@github.com] 
Sent: Thursday, October 12, 2017 12:35 AM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Mention <mention@noreply.github.com>
Subject: Re: [nvaccess/nvda] Object coordinates: It is now possible to hear object coordinates when using object navigation commands. re #2559 (#6078)

 

@josephsl <https://github.com/josephsl> : This has merge conflicts again. If fixed, I hope @michaelDCurran <https://github.com/michaeldcurran>  will be able to address this before 2018.1?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/6078#issuecomment-336046078> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AHgLkAqQ97_bR69OKbXt0iMuM70autTPks5srcEvgaJpZM4I3FmK> .

derekriemer commented:
what's wrong with indentation tones? We can rework some of it once speech refactor lands.

josephsl commented:
Hi, the issue being what to play first: indentation tone or coordinate beep. Thanks.

 

From: Derek Riemer [mailto:notifications@github.com] 
Sent: Thursday, October 12, 2017 7:38 AM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Mention <mention@noreply.github.com>
Subject: Re: [nvaccess/nvda] Object coordinates: It is now possible to hear object coordinates when using object navigation commands. re #2559 (#6078)

 

what's wrong with indentation tones? We can rework some of it once speech refactor lands.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/6078#issuecomment-336157402> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AHgLkDwXOEu8ZNpqbBYaN70B74aTzMirks5sriRJgaJpZM4I3FmK> .

josephsl committed:
Merged master

bdorer commented:
In my opinion I want to know intentation level and then object coordinates so I know the kind of object first and then where it is.

josephsl committed:
Merge branch 'master' into i2559

# Conflicts:
#	source/NVDAObjects/__init__.py
#	user_docs/en/userGuide.t2t

josephsl committed:
User guide: add anchors for object coordinate checkbox. re #2559

josephsl committed:
Update copyright years

Adriani90 commented:
This is amazing. This feature would help alot when creating power point presentations since NVDA can recognize the single elements on a slide as objects.

derekriemer commented:
waiting for speech refactor?

michaelDCurran commented:
Brightness detection is very slow on some Operating Systems. I would highly suggest not to support it for  play object coordinates. I.e. set it to false by default.  I also don't really think it is that necessary in this situation as the coordinate playd is an average position of the object, not a particular pixel anyway.

michaelDCurran commented:
Right now.  withobject coodinate tones and indentation tones both on at the same time, you will hear  object and then indentation.  I think this is the more logical order, though perhaps not the most practical order. For now though I'd be happy to take this as is, and revisit the order one speech refactor is in. 
In the long term I'd also like to see the ability to play multiple tones together so we can play the top left and bottom right at the same time, rather than just the center. But right now we don't have this capability.

michaelDCurran reviewed (changes requested):
Disable brightness detection for  playing object coordinates.

josephsl commented:
Hi, hmmm, I see. I guess it is redundant for now? Thanks.

 

From: Michael Curran <notifications@github.com> 
Sent: Sunday, April 15, 2018 4:59 PM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Mention <mention@noreply.github.com>
Subject: Re: [nvaccess/nvda] Object coordinates: It is now possible to hear object coordinates when using object navigation commands. re #2559 (#6078)

 

@michaelDCurran requested changes on this pull request.

Disable brightness detection for playing object coordinates.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/6078#pullrequestreview-112256867> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AHgLkJNzi28FZCAEF4gcPod7iECVPZNkks5to97DgaJpZM4I3FmK> .

josephsl committed:
Merge branch 'master' into i2559

josephsl committed:
Object coordinates: do not detect brightness, copyright year update. Re #2559.

Reviewed by Mick Curran (NV Access)L due to some issues, object brightness won't make sense when playing coordinates, so tell coordinates player function to ignore this.
Also updated copyright year.

michaelDCurran commented:
Looks good. But we'll  incubate this after multi category settings dialog goes to master as that will cause some conflicts.

josephsl committed:
Merge branch 'master' into i2559

# Conflicts:
#	source/gui/settingsDialogs.py

Adriani90 commented:
might be also related to #7599

josephsl commented:
Hi,

Queued for incubation after 2018.2 stable is released. Thanks.

josephsl committed:
Merge branch 'master' into i2559

josephsl committed:
Merge branch 'master' into i2559

josephsl committed:
Mouse handler: update copyright years

josephsl committed:
Mouse handler: set copyright header to that of NVDA 2018.2 (to be changed in 2018.3)

josephsl committed:
Mouse handler: restore removed imports.

josephsl committed:
Merge branch 'master' into i2559

josephsl commented:
Hi,

Resolved conflicts when attempting to incubate, traced to removed imports and copyright header differences.

Also, since no-one in Core (modules) are using mouseHandler.playAudioCoordinates, should it be deprecated? If so, I'd be happy to document that once merged into master.

Thanks.

michaelDCurran commented:
@josephsl can you provide an entry that can be used in changes.t2t please?

josephsl commented:
Hi,

A possible text:

NVDA will play a tone to indicate where the control is located on screen.

Thanks.

leonardder commented on source/NVDAObjects/__init__.py:
```diff
@@ -979,6 +980,8 @@ def event_becomeNavigatorObject(self, isFocus=False):
 		if braille.handler.shouldAutoTether and isFocus:
 			return
```
@josephsl: Could you please look at this early return? It looks like your code is going to play object coordinates for focus events when auto tethering is off. You probably want to remove the early return and change it like this:

```
		if not isFocus:
			if not braille.handler.shouldAutoTether:
				braille.handler.handleReviewMove(shouldAutoTether=not isFocus)
			if config.conf["presentation"]["playObjectCoordinates"]:
				screenExplorer.playObjectCoordinates(self)
```

LeonarddeR reviewed (commented):
I'm sorry I haven't looked at this earlier, but it looks like there is a small little bug here.

josephsl commented:
Hi, I see. Does what you’re proposing actually work? If so, I need few hours to test this. Thanks.

 

From: Leonard de Ruijter <notifications@github.com> 
Sent: Tuesday, June 26, 2018 11:45 PM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Mention <mention@noreply.github.com>
Subject: Re: [nvaccess/nvda] Object coordinates: It is now possible to hear object coordinates when using object navigation commands. re #2559 (#6078)

 

@leonardder commented on this pull request.

I'm sorry I haven't looked at this earlier, but it looks like there is a small little bug here.

  _____  

In source/NVDAObjects/__init__.py <https://github.com/nvaccess/nvda/pull/6078#discussion_r198381758> :

> @@ -979,6 +980,8 @@ def event_becomeNavigatorObject(self, isFocus=False):
               if braille.handler.shouldAutoTether and isFocus:
                       return

@josephsl <https://github.com/josephsl> : Could you please look at this early return? It looks like your code is going to play object coordinates for focus events when auto tethering is off. You probably want to remove the early return and change it like this:

               if not isFocus:
                       if not braille.handler.shouldAutoTether:
                               braille.handler.handleReviewMove(shouldAutoTether=not isFocus)
                       if config.conf["presentation"]["playObjectCoordinates"]:
                               screenExplorer.playObjectCoordinates(self)

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/6078#pullrequestreview-132301460> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AHgLkPu8P_GbIMu8Xp2G77FzgNMvlDi4ks5uAynkgaJpZM4I3FmK> .

josephsl commented:
Hi,

@LeonardDer: confirmed the bug. Because this will apply everywhere, I modified the event body to play the tone first if it's okay with you.

Thanks.

josephsl commented:
Hi,

A high severity bug found: if obj.location is None, Python raises TypeError. Relevant fix is part of latest i2559 branch commits. Thanks.

leonardder commented:
Has this ever been properly incubated into next? I just merged master 
into next to incorporate the most recent master changes (including 
translation updates), and there were code updates regarding this pr. And 
indeed, snapshot next-15427 does not contain the possibility to enable 
object coordinate tones.

michaelDCurran commented:
Oh... I thought it had. Let me look into this.

leonardder commented:
Please note that I had a quick run over the code and that I still have some concerns with the current implementation, I will review as soon as possible after the new pr is filed.
